apis/apiextensions/v1/xrd_webhooks.go

```diff
@@ -35,49 +36,55 @@ const (
 	errConversionWebhookConfigRequired = "spec.conversion.webhook is required when spec.conversion.strategy is 'Webhook'"
 )
 
+// NOTE(negz): We only validate updates because we're only using the validation
+// webhook to enforce a few immutable fields. We should look into using CEL per
+// https://github.com/crossplane/crossplane/issues/4128 instead.
+
 // +kubebuilder:webhook:verbs=update,path=/validate-apiextensions-crossplane-io-v1-compositeresourcedefinition,mutating=false,failurePolicy=fail,groups=apiextensions.crossplane.io,resources=compositeresourcedefinitions,versions=v1,name=compositeresourcedefinitions.apiextensions.crossplane.io,sideEffects=None,admissionReviewVersions=v1
 
 // ValidateCreate is run for creation actions.
-func (in *CompositeResourceDefinition) ValidateCreate() error {
+func (in *CompositeResourceDefinition) ValidateCreate() (admission.Warnings, error) {
+	// TODO(negz): Does this code ever get exercised in reality? We don't
+	// register the update verb when we generate a configuration above.
 	if c := in.Spec.Conversion; c != nil && c.Strategy == extv1.WebhookConverter && c.Webhook == nil {
-		return errors.New(errConversionWebhookConfigRequired)
+		return nil, errors.New(errConversionWebhookConfigRequired)
 	}
-	return nil
+	return nil, nil
 }
 
 // ValidateUpdate is run for update actions.
-func (in *CompositeResourceDefinition) ValidateUpdate(old runtime.Object) error {
+func (in *CompositeResourceDefinition) ValidateUpdate(old runtime.Object) (admission.Warnings, error) {
 	oldObj, ok := old.(*CompositeResourceDefinition)
 	if !ok {
-		return errors.New(errUnexpectedType)
+		return nil, errors.New(errUnexpectedType)
 	}
 	switch {
 	case in.Spec.Group != oldObj.Spec.Group:
-		return errors.New(errGroupImmutable)
+		return nil, errors.New(errGroupImmutable)
 	case in.Spec.Names.Plural != oldObj.Spec.Names.Plural:
-		return errors.New(errPluralImmutable)
+		return nil, errors.New(errPluralImmutable)
 	case in.Spec.Names.Kind != oldObj.Spec.Names.Kind:
-		return errors.New(errKindImmutable)
+		return nil, errors.New(errKindImmutable)
 	}
 	if in.Spec.ClaimNames != nil && oldObj.Spec.ClaimNames != nil {
 		switch {
 		case in.Spec.ClaimNames.Plural != oldObj.Spec.ClaimNames.Plural:
-			return errors.New(errClaimPluralImmutable)
+			return nil, errors.New(errClaimPluralImmutable)
 		case in.Spec.ClaimNames.Kind != oldObj.Spec.ClaimNames.Kind:
-			return errors.New(errClaimKindImmutable)
+			return nil, errors.New(errClaimKindImmutable)
 		}
 	}
-	return nil
+	return nil, nil
 }
 
 // ValidateDelete is run for delete actions.
-func (in *CompositeResourceDefinition) ValidateDelete() error {
-	return nil
+func (in *CompositeResourceDefinition) ValidateDelete() (admission.Warnings, error) {
+	return nil, nil
 }
 
 // SetupWebhookWithManager sets up webhook with manager.
-func (in *CompositeResourceDefinition) SetupWebhookWithManager(mgr ctrl.Manager) error {
+func SetupWebhookWithManager(mgr ctrl.Manager) error {
```

Comment:
> Do you expect to add here the code to setup al the webhooks for all resources in this package in one go?

Eventually - that's would be more similar to what we do with controllers (per the TODO in `core.go`).

Largely my thinking here was that where it's called:

```go
apiextensionsv1.SetupWebhookWithManager(mgr)
```

Seems a lot simpler (and perhaps more idiomatic?) than the approach where this is a method on a type:

```go
(&apiextensionsv1.CompositeResourceDefinition{}).SetupWebhookWithManager(mgr)
```

To me, it's a little confusing/unintuitive that calling Setup on one (empty) instance of a type is registering the entire type with the controller manager. Admittedly that's still happening with my update - it's just more in line with how we do every other `For(T)` call in the codebase, and IMO reads more like "setup a webook for this type" rather than "setup a webhook for this instance of this type".

Reply:
I agree it's not too nice, but I got used to it probably, it's a pretty common pattern: https://grep.app/search?q=Setupwebhookwithmanager

Reply:
Yeah it's not fundamentally wrong, and is what controller-runtime scaffolds I think. We diverged from that scaffolding with our `Setup` (for controller) pattern so I'd like to keep consistent.